src/coreComponents/constitutive/fluid/CubicEOSPhaseModel.hpp

```diff
@@ -0,0 +1,458 @@
+                            real64 & aMixtureCoefficient,
+                            real64 & bMixtureCoefficient )
+{
+  stackArray1d< real64, maxNumComps > m( numComps );
+  for( integer ic = 0; ic < numComps; ++ic )
+  {
+    m[ic] = EOS_TYPE::evaluate( acentricFactor[ic] );
+  }
+
+  // mixture coefficients
+  for( integer ic = 0; ic < numComps; ++ic )
+  {
+    aPureCoefficient[ic] = EOS_TYPE::omegaA * criticalTemperature[ic] * criticalTemperature[ic] * pressure
+                           / ( criticalPressure[ic] * temperature * temperature ) * pow( 1.0 + m[ic] * ( 1.0 - sqrt( temperature / criticalTemperature[ic] ) ), 2.0 );
+    bPureCoefficient[ic] = EOS_TYPE::omegaB * criticalTemperature[ic] * pressure / ( criticalPressure[ic] * temperature );
+  }
+
+  aMixtureCoefficient = 0.0;
+  bMixtureCoefficient = 0.0;
+  for( integer ic = 0; ic < numComps; ++ic )
+  {
+    for( integer jc = 0; jc < numComps; ++jc )
+    {
+      aMixtureCoefficient += ( composition[ic] * composition[jc] * ( 1.0 - binaryInteractionCoefficients ) * sqrt( aPureCoefficient[ic] * aPureCoefficient[jc] ) );
+    }
+    bMixtureCoefficient += composition[ic] * bPureCoefficient[ic];
+  }
+}
+
+template< typename EOS_TYPE >
+GEOS_HOST_DEVICE
+void
+CubicEOSPhaseModel< EOS_TYPE >::
+computeCompressibilityFactor( integer const numComps,
+                              arrayView1d< real64 const > const composition,
+                              real64 const & binaryInteractionCoefficients,
+                              arraySlice1d< real64 const > const aPureCoefficient,
+                              arraySlice1d< real64 const > const bPureCoefficient,
+                              real64 const & aMixtureCoefficient,
+                              real64 const & bMixtureCoefficient,
+                              real64 & compressibilityFactor )
+{
+  // a Z3 + b Z2 + cZ + d = 0
+  real64 const a = 1.0;
+  real64 const b = ( EOS_TYPE::delta1 + EOS_TYPE::delta2 - 1.0 ) * bMixtureCoefficient - 1.0;
+  real64 const c = aMixtureCoefficient + EOS_TYPE::delta1 * EOS_TYPE::delta2 * bMixtureCoefficient * bMixtureCoefficient
+                   - ( EOS_TYPE::delta1 + EOS_TYPE::delta2 ) * bMixtureCoefficient * ( bMixtureCoefficient + 1.0 );
+  real64 const d = -( aMixtureCoefficient * bMixtureCoefficient
+                      + EOS_TYPE::delta1 * EOS_TYPE::delta2 * bMixtureCoefficient * bMixtureCoefficient * ( bMixtureCoefficient + 1.0 ) );
+
+  real64 roots[3]{};
+  integer numRoots = 0;
+  solveCubicPolynomial( a, b, c, d, roots, numRoots );
+
+  if( numRoots == 1 )
+  {
+    compressibilityFactor = roots[0];
+  }
+  else
+  {
+    real64 zMin =  LvArray::NumericLimits< real64 >::max;
+    real64 zMax = -LvArray::NumericLimits< real64 >::max;
+
+    for( integer i = 0; i < 3; ++i )
+    {
+      // skip unphysical roots
+      if( roots[i] > bMixtureCoefficient )
+      {
+        // choose the root according to Gibbs' free energy minimization
+        if( zMin < roots[i] )
```

Comment:
The signs are reversed here I think.

Reply:
I meant shouldn't it be `if (zMin > roots[i])` and similarly for `zMax`.

Reply:
Yes!!! Thanks!